### ISSUE TITLE
eBPF scalability improvement

### ISSUE THREAD

**Issue by rootfs** [sustainable-computing-io/kepler#668]
- [ ] #665 
- [x] #671
- [x] tracking ebpf performance stats, thanks to [this ](https://ebpf.io/summit-2020-slides/eBPF_Summit_2020-Lightning-Bryce_Kahle-How_and_When_You_Should_Measure_CPU_Overhead_of_eBPF_Programs.pdf)

**Comment by rootfs** [comment]
after enabling bpf stats, this is kepler's stats
```console
# echo 1 > /proc/sys/kernel/bpf_stats_enabled
# bpftool prog show
....
1349: kprobe  name finish_task_swi  tag ffe5f649866774cd  gpl run_time_ns 1759368124 run_cnt 1080641
        loaded_at 2023-05-02T11:13:31-0400  uid 0
        xlated 2896B  jited 1620B  memlock 4096B  map_ids 1863,1862,1861,1860,1859,1858,1857,1856,1855,1854,1864
        btf_id 141
        pids kepler(1741884)
1350: tracepoint  name irq__softirq_en  tag c800d781274a6585  gpl run_time_ns 135136570 run_cnt 353914
        loaded_at 2023-05-02T11:13:31-0400  uid 0
        xlated 168B  jited 104B  memlock 4096B  map_ids 1864
        btf_id 141
        pids kepler(1741884)
```

**Comment by jichenjc** [comment]
>>>tracking ebpf performance stats, thanks to this 

I am interested in this , any proposed to do for this? e.g create a profile option so we can report such by turn on/off switch?

**Comment by rootfs** [comment]
@jichenjc we can use the ebpf run_time_ns and run_cnt to investigate how to reduce ebpf overhead and/or add more ebpf probes.

**Comment by jichenjc** [comment]
ok, let me check what I can do here..

**Comment by rootfs** [comment]
05/09 meeting:
- latency in rt kernels: whether disabling ebpf or disabling cgroup is more efficient and accurate. Choices should be supported by config options.
- cpu time is fine, but not as good as other perf counters per @marceloamaral research.

**Comment by rootfs** [comment]
runc vs containerd collect cgroup

**Comment by eklee15** [comment]
Some suggestions: 
1. Have we complied Kepler with rt-kernel and tested it?
2. Can we lower the priority of Kepler job and test it?

**Comment by mcalman** [comment]
I have begun investigating how we can improve the overhead of collecting cgroup metrics. Here are some initial results comparing the use of the containerd and runc libraries, as well as directly reading the cgroup files. These benchmarks are for v1 cgroups. I am working on replicating this for v2 groups and also examining CPU utilization.

![Image](https://github.com/sustainable-computing-io/kepler/assets/68564154/6e0516f6-8568-49ca-8388-ac233cb149de)

**Comment by marceloamaral** [comment]
Great Work @mcalman! Very interesting that containerd has better performance than runC.
Can you also measure the overhead, CPU and memory utilization?

**Comment by rootfs** [comment]
bpf latency is to be fixed by #823

**Comment by rootfs** [comment]
#823 merged, closing
